// @electron-forge/core/dist/util/publish-state.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

require("source-map-support/register");

var _crypto = _interopRequireDefault(require("crypto"));

var _fsExtra = _interopRequireDefault(require("fs-extra"));

var _path = _interopRequireDefault(require("path"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

const EXTENSION = '.forge.publish';

class PublishState {
  static async loadFromDirectory(directory, rootDir) {
    if (!(await _fsExtra.default.pathExists(directory))) {
      throw new Error(`Attempted to load publish state from a missing directory: ${directory}`);
    }

    const publishes = [];

    for (const dirName of await _fsExtra.default.readdir(directory)) {
      const subDir = _path.default.resolve(directory, dirName);

      const states = [];

      if ((await _fsExtra.default.stat(subDir)).isDirectory()) {
        const filePaths = (await _fsExtra.default.readdir(subDir)).filter(fileName => fileName.endsWith(EXTENSION)).map(fileName => _path.default.resolve(subDir, fileName));

        for (const filePath of filePaths) {
          const state = new PublishState(filePath);
          await state.load(); // eslint-disable-next-line max-len

          state.state.artifacts = state.state.artifacts.map(artifactPath => _path.default.resolve(rootDir, artifactPath));
          states.push(state);
        }
      }

      publishes.push(states);
    }

    return publishes;
  }

  static async saveToDirectory(directory, artifacts, rootDir) {
    const id = _crypto.default.createHash('SHA256').update(JSON.stringify(artifacts)).digest('hex');

    for (const artifact of artifacts) {
      // eslint-disable-next-line max-len
      artifact.artifacts = artifact.artifacts.map(artifactPath => _path.default.relative(rootDir, artifactPath));
      const publishState = new PublishState(_path.default.resolve(directory, id, 'null'), false);
      publishState.state = artifact;
      await publishState.saveToDisk();
    }
  }

  constructor(filePath, hasHash = true) {
    _defineProperty(this, "dir", void 0);

    _defineProperty(this, "path", void 0);

    _defineProperty(this, "hasHash", void 0);

    _defineProperty(this, "state", {});

    this.dir = _path.default.dirname(filePath);
    this.path = filePath;
    this.hasHash = hasHash;
  }

  generateHash() {
    const content = JSON.stringify(this.state || {});
    return _crypto.default.createHash('SHA256').update(content).digest('hex');
  }

  async load() {
    this.state = await _fsExtra.default.readJson(this.path);
  }

  async saveToDisk() {
    if (!this.hasHash) {
      this.path = _path.default.resolve(this.dir, `${this.generateHash()}${EXTENSION}`);
      this.hasHash = true;
    }

    await _fsExtra.default.mkdirs(_path.default.dirname(this.path));
    await _fsExtra.default.writeJson(this.path, this.state);
  }

}

exports.default = PublishState;
